### PR TITLE
Laboratory work #2, Yefimenko Natalya - 23FPL2

### PR DESCRIPTION
**Pull request by ShrubberyBush**

### CODE REVIEW

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -23,13 +20,14 @@ def tokenize(text: str) -> list[str] | None:
...
     """
+
     if not isinstance(text, str):
         return None
-    tokenized_list = []
-    for token in text.lower():
-        if token.isalpha():
-            tokenized_list.append(token)
-    return tokenized_list
+    tokens = []
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:26]: revert all changes in lab_1

File: lab_1_classify_by_unigrams/start.py [hunk 2]
```diff
@@ -2,9 +2,9 @@
 Language detection starter
 """
 # pylint:disable=too-many-locals, unused-argument, unused-variable
-from lab_1_classify_by_unigrams.main import (collect_profiles, create_language_profile,
-                                             detect_language, detect_language_advanced,
-                                             print_report, tokenize)
+
+from lab_1_classify_by_unigrams.main import (calculate_frequencies, calculate_mse, compare_profiles,
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:6]: the same

File: lab_2_retrieval_w_bm25/start.py [hunk 3]
```diff
@@ -1,7 +1,10 @@
 """
 Laboratory Work #2 starter
 """
-# pylint:disable=too-many-locals, unused-argument, unused-variable, too-many-branches, too-many-statements, duplicate-code
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/start.py:4]: revert these changes

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -19,6 +19,20 @@ def tokenize(text: str) -> list[str] | None:
     In case of corrupt input arguments, None is returned.
     """
 
+    if not isinstance(text, str):
+        return None
+    tokens = []
+    word = []
+    for token in text.lower():
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:26]: too rude, maybe there are some ways to split text into tokens?

**Review by EIzharikov — COMMENTED** [review]
You've started! This is already good, but you should hurry up! I wish you good luck

File: lab_2_retrieval_w_bm25/start.py (hunk 3 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/start.py:4]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:26]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:26]: Done I guess

File: lab_1_classify_by_unigrams/start.py (hunk 2 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/start.py:6]: Done I guess..? x2

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py]: no, you can check ```files changed ``` here
image

File: lab_1_classify_by_unigrams/start.py (hunk 2 above)
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/start.py]: nope

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -19,6 +21,22 @@ def tokenize(text: str) -> list[str] | None:
     In case of corrupt input arguments, None is returned.
     """
 
+    if not isinstance(text, str):
+        return None
+    tokens = []
+    for token in text.lower().split():
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:27]: here two loops, it is too much. Try to come up with a new way to tokenize.

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -48,6 +72,15 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
     In case of corrupt input arguments, None is returned.
     """
 
+    if not isinstance(documents, list) or not all(isinstance(doc, list) for doc in documents):
+        return None
+    vocabulary = set()
+    for doc in documents:
+        if not all(isinstance(token, str) for token in doc):
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:79]: check it earlier

File: lab_2_retrieval_w_bm25/main.py [hunk 7]
```diff
@@ -63,6 +96,19 @@ def calculate_tf(vocab: list[str], document_tokens: list[str]) -> dict[str, floa
     In case of corrupt input arguments, None is returned.
     """
 
+    if not isinstance(vocab, list) or not isinstance(document_tokens, list):
+        return None
+    if not all(isinstance(word, str) for word in vocab + document_tokens):
+        return None
+    tf_dict = {word: 0 for word in vocab}
+    doc_length = len(document_tokens)
+    if doc_length == 0:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:105]: this if is redundant

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -295,4 +295,4 @@ def print_report(detections: list[tuple[str, float]]) -> None:
     In case of corrupt input arguments, None is returned
     """
     for detection in detections:
-        print(f'{detection[0]}: MSE {detection[-1]:.5f}')
+        print(f'{detection[0]}: MSE {detection[-1]:.5f}')
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:298]: revert

File: lab_1_classify_by_unigrams/start.py [hunk 9]
```diff
@@ -41,4 +41,4 @@ def main() -> None:
 
 
 if __name__ == "__main__":
-    main()
+    main()
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:44]: revert

**Review by EIzharikov — COMMENTED** [review]
You've made some progress, but it's still progress! I wish you good luck in further work.

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py]: I think I've figured it out ._.

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py]: Done

File: lab_1_classify_by_unigrams/start.py (hunk 9 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/start.py]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:27]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 6 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:79]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 7 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:105]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (EIzharikov) [reply lab_2_retrieval_w_bm25/main.py:28]: what done? it is not rude now. Try to use .split() function

File: lab_2_retrieval_w_bm25/main.py [hunk 10]
```diff
@@ -63,6 +98,25 @@ def calculate_tf(vocab: list[str], document_tokens: list[str]) -> dict[str, floa
     In case of corrupt input arguments, None is returned.
     """
 
+    if not vocab or not document_tokens:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:101]: join checks

File: lab_2_retrieval_w_bm25/main.py [hunk 11]
```diff
@@ -78,6 +132,25 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
     In case of corrupt input arguments, None is returned.
     """
 
+    if not vocab or not documents:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:135]: join checks

File: lab_2_retrieval_w_bm25/main.py [hunk 12]
```diff
@@ -93,6 +166,19 @@ def calculate_tf_idf(tf: dict[str, float], idf: dict[str, float]) -> dict[str, f
     In case of corrupt input arguments, None is returned.
     """
 
+    if not tf or not idf:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:169]: join checks

File: lab_2_retrieval_w_bm25/main.py [hunk 13]
```diff
@@ -121,6 +207,35 @@ def calculate_bm25(
...
+    if ((not all(isinstance(value, str) for value in vocab)) or
+            (not all(isinstance(value, str) for value in document))
+            or (not all(isinstance(key, str) and isinstance(value, float) for
+                        key, value in idf_document.items()))):
+        return None
+    if ((not isinstance(k1, float)) or (not isinstance(b, float))
+            or ((not isinstance(avg_doc_len, float)) or
+                (not isinstance(doc_len, int)) or (isinstance(doc_len, bool)))):
+        return None
+    if avg_doc_len == 0:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:224]: what if we will have negative number?

File: lab_2_retrieval_w_bm25/main.py [hunk 14]
```diff
@@ -121,6 +207,35 @@ def calculate_bm25(
...
+        return None
+    bm25 = {}
+    for token in vocab:
+        if token not in idf_document:
+            bm25[token] = 0.0
+        else:
+            token_count = document.count(token)
+            denominator = token_count + k1 * (1 - b + b * (doc_len / avg_doc_len))
+            if denominator == 0:
+                bm25[token] = 0.0
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:234]: when it can be?

File: lab_2_retrieval_w_bm25/main.py [hunk 15]
```diff
@@ -139,6 +254,34 @@ def rank_documents(
     In case of corrupt input arguments, None is returned.
     """
 
+    if (not isinstance(indexes, list) or
+            not all(isinstance(el, dict) and
+                    all(isinstance(key, str)
+                        and isinstance(value, float)
+                        for key, value in el.items()) for el in indexes)):
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:261]: what is el?

File: lab_2_retrieval_w_bm25/main.py [hunk 16]
```diff
@@ -139,6 +254,34 @@ def rank_documents(
...
+        return None
+    if (not isinstance(stopwords, list) or not all(isinstance(el, str) for el in stopwords)
+            or not isinstance(query, str)):
+        return None
+    if not indexes or not query:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:266]: join checks

**Review by EIzharikov — COMMENTED** [review]
fix all comments, make CI green and be ready for the next review. Good progress, I wish you good luck

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:28]: I used it and it didn't work, now it's fixed

File: lab_2_retrieval_w_bm25/main.py (hunk 10 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:101]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 11 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:135]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 12 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:169]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 13 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:224]: Fixed

File: lab_2_retrieval_w_bm25/main.py (hunk 14 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:234]: If k1 = 0 and token_count = 0

File: lab_2_retrieval_w_bm25/main.py (hunk 15 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:261]: Element from indexes

File: lab_2_retrieval_w_bm25/main.py (hunk 16 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:266]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 15 above)
Comment (EIzharikov) [reply lab_2_retrieval_w_bm25/main.py:255]: rename to index

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
Comment (EIzharikov) [reply lab_2_retrieval_w_bm25/main.py:27]: where?

**Review by EIzharikov — COMMENTED** [review]
So, you have unfixed comments. Make CI green, answer&fix all my comments and be ready for Friday. Good luck.

File: lab_2_retrieval_w_bm25/main.py (hunk 15 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:255]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
Comment (ShrubberyBush) [reply lab_2_retrieval_w_bm25/main.py:27]: Done..?
